R/backport_linter.R

```diff
@@ -0,0 +1,83 @@
+#' @describeIn linters that checks for usage of unavailable functions. Not reliable for testing r-devel dependencies.
+#' @export
+backport_linter <- function(r_version = "4.0.3") {
+  function(source_file) {
+    if (is.null(source_file$xml_parsed_content) || r_version >= "r-devel") return(list())
+
+    xml <- source_file$xml_parsed_content
+
+    #browser()
+
+    names_xpath <- "//*[self::SYMBOL or self::SYMBOL_FUNCTION_CALL]"
+    all_names_nodes <- xml2::xml_find_all(xml, names_xpath)
+    all_names <- xml2::xml_text(all_names_nodes)
+
+    # guaranteed to include 1 by early return above; which.min fails if all TRUE (handled by nomatch)
+    needs_backport_names <- backports[1:(match(FALSE, r_version < names(backports), nomatch = length(backports)) - 1L)]
+
+    # not sapply, which may over-simplify -- cbind makes sure apply works
+    needs_backport <- do.call(cbind, lapply(needs_backport_names, function(nm) all_names %in% nm))
```

Comment:
`do.call(cbind, lapply(...))` looks like `vapply()` might be the way to go.
For atomic length `n > 1` results, `vapply` cbinds the results into a matrix.

Reply:
The issue is with `n=1` (`length(all_names)=1`), where a vector is returned and then `apply` fails. `cbind` guarantees we have a matrix (I tried `sapply` first and got the same issue, I'll add `vapply` to the comment)

Reply:
Thanks for clarifying :)
An alternative would be `dim(needs_backport) <- c(length(all_names), length(needs_backport_names))` after `vapply`. But I don't mind since the number of columns is bounded by the number of backports - the `do.call(cbind, ...)` overhead is at least only constant and independent of user code.

Reply:
Can also just `cbind` or `rbind` the `vapply` result to get matrix, but I figured it wouldn't improve readability. Hadn't thought of scalability but yes, the number of rows comes from the code size while we control the number of columns.